pkg/destroy/openstack/openstack.go

```diff
@@ -422,6 +452,32 @@ func deletePorts(opts *clientconfig.ClientOpts, listOpts ports.ListOpts, logger
 				}
 			}
 
+			// If there is a security group created by cloud-provider-openstack we should find it and delete it.
+			// We'll look through the ones on each of the ports and attempt to remove it from the port and delete it.
+			// Most of the time it's a conflict, but last port should be guaranteed to allow deletion.
+			// TODO(dulek): Currently this is the only way to do it and if delete fails there's no way to get back to
+			//              that SG. This is bad and we should make groups created by CPO tagged by cluster ID ASAP.
+			assignedSGs := port.SecurityGroups
+			ports.Update(conn, port.ID, ports.UpdateOpts{
+				SecurityGroups: &[]string{}, // We can just detach all, we're deleting this port anyway.
+			})
+			for _, groupID := range assignedSGs {
+				if group, ok := sgByID[groupID]; ok {
+					if cloudProviderSGNameRegexp.MatchString(group.Name) {
+						logger.Debugf("Deleting cloud-provider-openstack SG %q", groupID)
+						err := sg.Delete(conn, groupID).ExtractErr()
+						var err404 gophercloud.ErrDefault404
+						var err409 gophercloud.ErrDefault409
+						if errors.As(err, &err404) {
```

Comment:
can't we clean the sgID from the map right after the successful operation of deleting the SG (in other words when there is no error)?

Reply:
Ah, okay, you're right.